### ISSUE TITLE
Add support for CircleCI tokens

### ISSUE THREAD

**Issue by di** [sigstore/fulcio#591]
**Description**

Similar to #243, CircleCI supports OIDC tokens: https://circleci.com/docs/2.0/openid-connect-tokens/

Also similar to #243, the `aud` is not customizable: https://circleci.canny.io/cloud-feature-requests/p/customizable-audience-claim-in-oidc-tokens

```
$ curl https://gitlab.com/.well-known/openid-configuration | jq
```
```json
{
  "issuer": "https://gitlab.com",
  "authorization_endpoint": "https://gitlab.com/oauth/authorize",
  "token_endpoint": "https://gitlab.com/oauth/token",
  "revocation_endpoint": "https://gitlab.com/oauth/revoke",
  "introspection_endpoint": "https://gitlab.com/oauth/introspect",
  "userinfo_endpoint": "https://gitlab.com/oauth/userinfo",
  "jwks_uri": "https://gitlab.com/oauth/discovery/keys",
  "scopes_supported": [
    "api",
    "read_api",
    "read_user",
    "read_repository",
    "write_repository",
    "read_registry",
    "write_registry",
    "sudo",
    "openid",
    "profile",
    "email"
  ],
  "response_types_supported": [
    "code"
  ],
  "response_modes_supported": [
    "query",
    "fragment"
  ],
  "grant_types_supported": [
    "authorization_code",
    "password",
    "client_credentials",
    "refresh_token"
  ],
  "token_endpoint_auth_methods_supported": [
    "client_secret_basic",
    "client_secret_post"
  ],
  "subject_types_supported": [
    "public"
  ],
  "id_token_signing_alg_values_supported": [
    "RS256"
  ],
  "claim_types_supported": [
    "normal"
  ],
  "claims_supported": [
    "iss",
    "sub",
    "aud",
    "exp",
    "iat",
    "sub_legacy",
    "name",
    "nickname",
    "email",
    "email_verified",
    "website",
    "profile",
    "picture",
    "groups",
    "groups_direct",
    "https://gitlab.org/claims/groups/owner",
    "https://gitlab.org/claims/groups/maintainer",
    "https://gitlab.org/claims/groups/developer"
  ]
}
```

**Comment by nsmith5** [comment]
@zalimeni I see on the [Circle roadmap](https://circleci.com/product-roadmap/) that API tokens getting auto injected into builds will be a think in the future. Is there any change that these tokens will be usable to grab OIDC tokens with custom claims. IIRC this is how it works in Github actions. There is a token with a default `aud` but you can request a new one with a different audience inside you build

**Comment by fproulx-boostsecurity** [comment]
They shipped. so can this be added soon ?

**Comment by nsmith5** [comment]
No, they don't support customizable 'aud' yet afaik so we can't support in
Fulcio yet. See the attached feature request in canny for details

**Comment by nsmith5** [comment]
This work is now planned at CircleCI https://circleci.canny.io/cloud-feature-requests/p/customizable-audience-claim-in-oidc-tokens

**Comment by MarshallOfSound** [comment]
@di / @nsmith5 Looks like this is getting underway at Circle, wondering if the requirement from fulcio is specifically that `aud` is **exactly equal** to `sigstore` or per the JWT spec can it be an array of audiences where one of them is `sigstore`?

**Comment by nsmith5** [comment]
@MarshallOfSound Still marked as "planned" on Canny as it has been for a while...

In any case, we use https://pkg.go.dev/github.com/coreos/go-oidc/v3/oidc#IDTokenVerifier.Verify to verify token which should be conformant to the [OIDC Spec](https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation)

From the spec section 3.1.3.7.  ID Token Validation:

> The ID Token MUST be rejected if the ID Token does not list the Client as a valid audience, **or if it contains additional audiences not trusted by the Client.**

So my assumption is that additional audiences will be rejected. Though when you look at the actually implementation, it seems to ignore additional audiences if it finds the the right one at least.

https://github.com/coreos/go-oidc/blob/v3.5.0/oidc/verify.go#L264

Soooo yeah this will work, but I'd caution CircleCI folks from implementing this way has it appears to be outside of OIDC conformance and still has the same security flaws as a fixed audience. If you can only add more and more audience claims the tokens still can't be narrowly scoped to a client basically..

**Comment by MarshallOfSound** [comment]
@nsmith5 There's definitely movement happening behind the scenes, I'm just trying to make sure that what's being done will be workable / supported by fulcio as that is the main reason we're pushing for customizable audience claims on CircleCI.

Sounds like the array-of-audiences approach although technically will work should probably not work in the future 😅 Thanks for the reference, super helpful 👍

**Comment by nsmith5** [comment]
😆 Yeah I'm honestly tempted to patch the OIDC library to be closer to the spec. I've you've got an leverage as a customer with them and can push them in a particular way, I'd definitely advise them towards making the implementation as close to the OIDC spec as possible. Better security IMO

**Comment by jerdog** [comment]
CircleCI has [just added](https://circleci.canny.io/cloud-feature-requests/p/customizable-audience-claim-in-oidc-tokens) customizable audience claims in OIDC tokens =)

https://circleci.com/docs/api/v2/index.html#tag/OIDC-Token-Management

**Comment by haydentherapper** [comment]
Fantastic! Next step would be to update the OID doc to map claims in the CircleCI identity token to the extensions in Fulcio certs.

**Comment by nsmith5** [comment]
@jerdog I'm a little tentative about accepting tokens from CircleCI. Unfortunately the implementation only allows adding audiences to the one token issued per-build. If someone configures more than one client / audience in their build it would be clear to Fulcio whether or not the token has come from the pipeline or the _other_ audience / client. I'll copy my example [from the Canny](https://circleci.canny.io/cloud-feature-requests/p/customizable-audience-claim-in-oidc-tokens) here:

> As an example, imagine you want to authenticate to service A and service B from your build. In this case you'd set the audience to something like [ "service A", "service B"] on that project and each service should see that it is in the audience list and accept the token. The problem is that if service A gets compromised or is acting maliciously, it can take the token it received from your build and use it to authenticate to service B.

The design makes is a bit more likely that Fulcio could issue code signing certificates to the wrong client and that is a concern to me. 

This isn't just a concern for Sigstore, but rather all OIDC clients consuming CircleCI tokens with a list of audiences and is very similar to the original security concern described in the canny issue -- A single hard coded audience has the same security properties that a list of many audiences on one token has.

**Comment by jerdog** [comment]
I'll pass this on @nsmith5

**Comment by MarshallOfSound** [comment]
I've passed this on too, this design is actually why I was asked my questions a few weeks ago 😄 Hopefully we see the ability to generate multiple per-audience tokens so we can get this ball moving

**Comment by takeoutweight** [comment]
Hi! I'm Nathan from CircleCI. We're just starting to roll out the ability to configure the `aud` claim in OIDC tokens at a fine-grained level. It's used in CI like:

```sh
export SIG_TOKEN=$(circleci run oidc get --claims '{"aud": "sigstore"}')
```

All the other claims are as per the [$CIRCLE_OIDC_TOKEN_V2](https://circleci.com/docs/openid-connect-tokens/#format-of-the-openid-connect-id-token). We don't allow any other customization with this mechanism besides the "aud" claim.

Just checking in to see what next steps might look like for getting CircleCI set up as a supported identity provider. I see from [Mapping OIDC token claims to Fulcio OIDs](https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md#mapping-oidc-token-claims-to-fulcio-oids) there are a number of claims that can map to cert extensions for code signing. We don't currently supply all of these data points in our tokens so I'm guessing we'd be looking at expanding our support for code signing after becoming an IDP, as a follow up?

**Comment by fproulx-boostsecurity** [comment]
This is indeed officially documented in CircleCI's documentation https://circleci.com/docs/oidc-tokens-with-custom-claims/

So what's next? Add it to Fulcio, then in `cosign` ?

**Comment by haydentherapper** [comment]
Yep, if you take a look at the CodeFresh PR, that’s what you’ll need to do. Add the code and map as many token claims to extensions.

**Comment by meeech** [comment]
👋🏼 @haydentherapper what nice timing. I'm a someone from circleci, I just started looking at this yesterday for pypi (thank you responsive pypi security team who pointed me at the relevant threads!) 

I will look at the mapping tables / configs to understand what needs to happen. I assume this thread will be ok to use if I have questions?

**Comment by meeech** [comment]
perfect i'm in. thanks hayden.
